pkg/service/cluster/service.go

```diff
@@ -171,44 +174,96 @@ func (s *Service) discoverAndSetClusterHosts(ctx context.Context, c *Cluster) er
 	return errors.Wrap(s.setKnownHosts(c, knownHosts), "update known_hosts in SM DB")
 }
 
-func (s *Service) discoverClusterHosts(ctx context.Context, c *Cluster) ([]string, error) {
-	var contactPoints []string
+func (s *Service) discoverClusterHosts(ctx context.Context, c *Cluster) (knownHosts, liveHosts []string, err error) {
 	if c.Host != "" {
-		contactPoints = append(contactPoints, c.Host) // Go with the designated contact point first
+		knownHosts, liveHosts, err := s.discoverClusterHostUsingCoordinator(ctx, c, 5*time.Second, c.Host)
+		if err != nil {
+			s.logger.Error(ctx, "Couldn't discover hosts using stored coordinator host, proceeding with other known ones",
+				"coordinator-host", c.Host, "error", err)
+		} else {
+			return knownHosts, liveHosts, nil
+		}
 	} else {
 		s.logger.Error(ctx, "Missing --host flag. Using only previously discovered hosts instead", "cluster ID", c.ID)
 	}
-	contactPoints = append(contactPoints, c.KnownHosts...) // In case it failed, try to contact previously discovered hosts
+	if len(c.KnownHosts) < 1 {
+		return nil, nil, ErrNoValidKnownHost
+	}
 
-	for _, cp := range contactPoints {
-		if cp == "" {
-			s.logger.Error(ctx, "Empty contact point", "cluster ID", c.ID, "contact points", contactPoints)
-			continue
-		}
+	wg := sync.WaitGroup{}
+	type hostsTuple struct {
+		live, known []string
+	}
+	result := make(chan hostsTuple, len(c.KnownHosts))
+	done := make(chan struct{})
+	discoverContext, discoverCancel := context.WithCancel(ctx)
+	defer discoverCancel()
 
-		config := scyllaclient.DefaultConfigWithTimeout(s.timeoutConfig)
-		if c.Port != 0 {
-			config.Port = fmt.Sprint(c.Port)
-		}
-		config.AuthToken = c.AuthToken
-		config.Hosts = []string{cp}
+	for _, cp := range c.KnownHosts {
+		wg.Add(1)
 
-		client, err := scyllaclient.NewClient(config, s.logger.Named("client"))
-		if err != nil {
-			s.logger.Error(ctx, "Couldn't connect to contact point", "contact point", cp, "error", err)
-			continue
-		}
+		go func(host string) {
+			defer wg.Done()
 
-		knownHosts, err := s.discoverHosts(ctx, client)
-		logutil.LogOnError(ctx, s.logger, client.Close, "Couldn't close scylla client")
-		if err != nil {
-			s.logger.Error(ctx, "Couldn't discover hosts", "host", cp, "error", err)
-			continue
+			knownHosts, liveHosts, err := s.discoverClusterHostUsingCoordinator(discoverContext, c, 5*time.Second, host)
+			if err != nil && !errors.Is(err, context.Canceled) {
+				s.logger.Error(ctx, "Couldn't discover hosts", "host", host, "error", err)
+				return
+			}
```

Comment:
What if the error is `context.Canceled`? Shouldn't we return as well?

Reply:
not necessary, but yes let's return

Reply:
Why is it not necessary? If we don't return, we still put some undefined values into the `result` chan, and the main function body reading from this chan will believe that everything is ok, when it's not.

Reply:
It's exactly "undefined" values, but just nil.
The main function body is expected to read to correct result already, so it would be put to the channel but never consumed.

Reply:
But if the parent context is cancelled, then it triggers child ones to be cancelled as well.
It means that the function would return "nil, nil, nil", what is wrong.

But now, after fixing this what you suggested it will work fine and the "ErrNoValidKnownHost" is going to be returned. Nice !

Reply:
That's a valid statement.
Replaced with `discoverContext.Err() != context.Canceled`
I want to avoid the situation where one of the nodes reported the live and known hosts. 
DiscoverContext is cancelled and the goroutinet logs this information.These logs are unnecessary. But would be good to know if the error reason is different.